argon2/src/block.rs

```diff
@@ -154,3 +154,43 @@ impl Zeroize for Block {
         self.0.zeroize();
     }
 }
+
+/// Custom implementation of `Box<[Block]>` until `Box::try_new_zeroed_slice` is stabilized.
+#[cfg(feature = "alloc")]
+pub(crate) struct Blocks {
+    p: core::ptr::NonNull<Block>,
+    len: usize,
+}
+
+#[cfg(feature = "alloc")]
+impl Blocks {
+    pub fn new(len: usize) -> Option<Self> {
+        use alloc::alloc::{Layout, alloc_zeroed};
+        use core::ptr::NonNull;
+
+        let layout = Layout::array::<Block>(len).ok()?;
+        // SAFETY: e use `alloc_zeroed` correctly
+        let p = unsafe { alloc_zeroed(layout) };
```

Comment:
This depends on the caller (which, to be fair, is limited to this crate) not passing `len == 0`.

Wouldn't it be better to explicitly assert that `layout.size() > 0`?

Reply:
Ah, indeed. `alloc_zeroed` requires for `Layout` to be non-zero. We probably can simply return `None` in such case instead of using a dangling pointer on zero length.